### ISSUE TITLE
phpmyadmin/sql-parser dependency has a 'viral license'

### ISSUE THREAD

**Issue by gisostallenberg** [larastan/larastan#1373]
- Larastan Version: v2.2.0
- `--level` used: N/A
- Pull request with failing test: N/A

### Description
Larastan uses `phpmyadmin/sql-parser` in `PhpMyAdminDataTypeToPhpTypeConverter` through the `SquashedMigrationHelper`, which is used in the `ModelPropertyExtension`. This package has a so called [Viral license](https://en.wikipedia.org/wiki/Viral_license) (GPL-2.0-or-later). Because of the 'viral' aspect of this, AFAIK this means that we are actually unable to use Larastan in a closed source project.
Would it be possible to 'easily' remove this dependency and would you be willing to do so?

I would really like to contribute if this is necessary. Could you point me to what's need to be changed?

**Comment by marijnvanderhorst** [comment]
I just came across this license dependency as well. I think that, because of the `phpmyadmin/sql-parser` dependency, larastan as a library is actually not allowed to be licensed under MIT. 

Still, as long as you use larastan as a development dependency only, it should be fine to use it in a commercial/closed-source project I believe. This, as the application itself is then not actually linked against the library and the library is not used at runtime.

I'm not an expert on this subject at all though so please do correct me if I'm wrong!

**Comment by ondrejmirtes** [comment]
Feel free to try this instead: https://github.com/SQLFTW/sqlftw

**Comment by williamdes** [comment]
Oh, I did not know our library was used here. Nothing we can do @ibennetch ?

**Comment by gisostallenberg** [comment]
@williamdes and @ibennetch cool to see this got your attention. I hope you can offer a solution at the license level, which would avoid changing this codebase.

**Comment by williamdes** [comment]
See https://github.com/phpmyadmin/sql-parser/issues/395

**Comment by alu-** [comment]
I've forked larastan and replaced phpmyadmin/sql-parser with iamcal/sql-parser. All tests pass and I've tested the changes against our repository, but I don't consider that a large enough data set. I'd really like others to try these temporary fix.
https://github.com/alu-/larastan/tree/eliminate-viral-licenses

To try this yourself you can modify composer.json to include:
```
"repositories": [
    {
        "type": "vcs",
        "url": "https://github.com/alu-/larastan"
    }
],
"require-dev": {
    "nunomaduro/larastan": "dev-eliminate-viral-licenses",
}
```

**Comment by jonnylink** [comment]
> @williamdes and @ibennetch cool to see this got your attention. I hope you can offer a solution at the license level, which would avoid changing this codebase.

Why was this even added? With all due respect (and I respect people working on this great package a lot), adding a new dependency without understanding its license seems like a mistake. Expecting it to change feels a bit too optimistic. 

Looks like someone already forked to fix this on their own. Any chance we can pull those changes into the main branch? 

For now I'm reverting to the deprecated package.

**Comment by RobertBoes** [comment]
It's been almost 2 years since this has been reported. I get that changing the `phpmyadmin/sql-parser` license would take a long time. But why hasn't Larastan taken any effort to resolve this in a different way?

AFAIK quite a few repositories are violating licenses because of it, as they would need to use the same license
- Larastan itself (as it's MIT licensed)
- https://github.com/nunomaduro/collision/blob/v8.x/LICENSE.md
- https://github.com/laravel/pint/blob/main/LICENSE.md
- 1200 packages on packagist, of which a lot use a MIT license https://packagist.org/packages/larastan/larastan/dependents?order_by=downloads

And yes, larastan in a closed source application would mean the source needs to be disclosed.

**Comment by williamdes** [comment]
Update from our side, LGPL/GPL dual licensing seems to be a path.
Awaiting news on Conservancy, I am sending them a ping email today.
Ref: https://github.com/phpmyadmin/sql-parser/issues/395#issuecomment-1479279462

Official issue: https://github.com/phpmyadmin/sql-parser/issues/448

**Comment by calebdw** [comment]
> And yes, larastan in a closed source application would mean the source needs to be disclosed.

This is not true. The majority of (if not all) the projects using Larastan are using it as a **dev** dependency for static/code analysis. This means that Larastan and its dependencies (including the GPL sql-parser package in question) are not included in the production build nor distributed as part of the production application thereby not invoking the viral aspect of the GPL's copyleft provisions.

**Comment by nunomaduro** [comment]
I’ve just seen this issue for the first time and feel that we need to do one of two things:

- Remove the feature that uses SQL parsing.
- Use a different SQL parsing tool for this feature (one of the MIT tools mentioned above).

I’m going to wait for @canvural’s decision—he is the main maintainer of this project at this point.

**Comment by calebdw** [comment]
I have a feature idea (https://github.com/larastan/larastan/discussions/2211) which would remove the needs for sql parsing and directly query the database (with caching of course) for the model schema

**Comment by nunomaduro** [comment]
Alright, I’ve decided to be proactive, and in light of what I read about this issue just an hour ago, and not knowing the availability of @canvural to address it, I’ve decided to tag new versions of both Larastan v2 and v3, effectively replacing `phpmyadmin/sql-parser` with `iamcal/sql-parser`.

- [Larastan v3.2.0](https://github.com/larastan/larastan/releases/tag/v3.2.0)
- [Larastan v2.10.0](https://github.com/larastan/larastan/releases/tag/v2.10.0)

In the future, if `phpmyadmin/sql-parser` changes to a license to something we can use, I am happy to revert this change. Thank you all for reporting this issue.

**Comment by alu-** [comment]
Cool. Thanks for changing to my idea of using iamcal/sql-parser. :)

**Comment by MauricioFauth** [comment]
> In the future, if phpmyadmin/sql-parser changes to a license to something we can use, I am happy to revert this change.

@nunomaduro @canvural We are willing to change the license to LGPL 2.1, which will allow Larastan to keep the MIT license. Would that work for you?

**Comment by calebdw** [comment]
@MauricioFauth, converting the License to LGPL would work and would be great for other projects that use the parser.

It seems there are issues with the new parser (https://github.com/larastan/larastan/issues/2230), so updating would allow us to revert back to the phpmyadmin parser

**Comment by tborealis** [comment]
We've had to pin to Larastan v3.1.0 as the new SQL parser is incompatible with PostgreSQL. As this is a backwards incompatible change, perhaps a major version increment is needed?

**Comment by tayeke** [comment]
> We've had to pin to Larastan v3.1.0 as the new SQL parser is incompatible with PostgreSQL. As this is a backwards incompatible change, perhaps a major version increment is needed?

I also cannot use v3.2.0 with PostgreSQL. All of the CHARACTER and BPCHAR types are throwing errors.

**Comment by nunodonato** [comment]
I'm just leaving this here as I was another victim of this change https://github.com/larastan/larastan/discussions/2300
Also pinned to `3.1.0`

**Comment by williamdes** [comment]
See https://github.com/larastan/larastan/pull/2388
And release notes https://github.com/larastan/larastan/releases/tag/v3.8.0

The phpMyAdmin sql-parser is back and optional

**Comment by gisostallenberg** [comment]
@canvural and @williamdes Nice solution to make it optional.

Thanks for the efforts you have put into this!
